# Importing Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [39]:
df1=pd.read_csv("matches.csv")

In [3]:
df1

id  season           city        date                        team1  \
0        1    2017      Hyderabad  2017-04-05          Sunrisers Hyderabad   
1        2    2017           Pune  2017-04-06               Mumbai Indians   
2        3    2017         Rajkot  2017-04-07                Gujarat Lions   
3        4    2017         Indore  2017-04-08       Rising Pune Supergiant   
4        5    2017      Bangalore  2017-04-08  Royal Challengers Bangalore   
..     ...     ...            ...         ...                          ...   
751  11347    2019         Mumbai    05/05/19        Kolkata Knight Riders   
752  11412    2019        Chennai    07/05/19          Chennai Super Kings   
753  11413    2019  Visakhapatnam    08/05/19          Sunrisers Hyderabad   
754  11414    2019  Visakhapatnam    10/05/19               Delhi Capitals   
755  11415    2019      Hyderabad    12/05/19               Mumbai Indians   

                           team2                  toss_winner toss_decision  \
0    Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1         Rising Pune Supergiant       Rising Pune Supergiant         field   
2          Kolkata Knight Riders        Kolkata Knight Riders         field   
3                Kings XI Punjab              Kings XI Punjab         field   
4               Delhi Daredevils  Royal Challengers Bangalore           bat   
..                           ...                          ...           ...   
751               Mumbai Indians               Mumbai Indians         field   
752               Mumbai Indians          Chennai Super Kings           bat   
753               Delhi Capitals               Delhi Capitals         field   
754          Chennai Super Kings          Chennai Super Kings         field   
755          Chennai Super Kings               Mumbai Indians           bat   

     result  dl_applied                       winner  win_by_runs  \
0    normal           0          Sunrisers Hyderabad           35   
1    normal           0       Rising Pune Supergiant            0   
2    normal           0        Kolkata Knight Riders            0   
3    normal           0              Kings XI Punjab            0   
4    normal           0  Royal Challengers Bangalore           15   
..      ...         ...                          ...          ...   
751  normal           0               Mumbai Indians            0   
752  normal           0               Mumbai Indians            0   
753  normal           0               Delhi Capitals            0   
754  normal           0          Chennai Super Kings            0   
755  normal           0               Mumbai Indians            1   

     win_by_wickets player_of_match  \
0                 0    Yuvraj Singh   
1                 7       SPD Smith   
2                10         CA Lynn   
3                 6      GJ Maxwell   
4                 0       KM Jadhav   
..              ...             ...   
751               9       HH Pandya   
752               6        AS Yadav   
753               2         RR Pant   
754               6    F du Plessis   
755               0       JJ Bumrah   

                                         venue         umpire1  \
0    Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1      Maharashtra Cricket Association Stadium  A Nand Kishore   
2       Saurashtra Cricket Association Stadium     Nitin Menon   
3                       Holkar Cricket Stadium    AK Chaudhary   
4                        M Chinnaswamy Stadium             NaN   
..                                         ...             ...   
751                           Wankhede Stadium   Nanda Kishore   
752                  M. A. Chidambaram Stadium     Nigel Llong   
753                           ACA-VDCA Stadium             NaN   
754                           ACA-VDCA Stadium   Sundaram Ravi   
755         Rajiv Gandhi Intl. Cricket Stadium     Nitin Menon   

            umpire2            

In [94]:
df2=pd.read_csv("deliveries.csv")

In [5]:
df2

match_id  inning         batting_team                 bowling_team  \
0              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
1              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
2              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
3              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
4              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
...          ...     ...                  ...                          ...   
179073     11415       2  Chennai Super Kings               Mumbai Indians   
179074     11415       2  Chennai Super Kings               Mumbai Indians   
179075     11415       2  Chennai Super Kings               Mumbai Indians   
179076     11415       2  Chennai Super Kings               Mumbai Indians   
179077     11415       2  Chennai Super Kings               Mumbai Indians   

        over  ball    batsman non_striker      bowler  is_super_over  ...  \
0          1     1  DA Warner    S Dhawan    TS Mills              0  ...   
1          1     2  DA Warner    S Dhawan    TS Mills              0  ...   
2          1     3  DA Warner    S Dhawan    TS Mills              0  ...   
3          1     4  DA Warner    S Dhawan    TS Mills              0  ...   
4          1     5  DA Warner    S Dhawan    TS Mills              0  ...   
...      ...   ...        ...         ...         ...            ...  ...   
179073    20     2  RA Jadeja   SR Watson  SL Malinga              0  ...   
179074    20     3  SR Watson   RA Jadeja  SL Malinga              0  ...   
179075    20     4  SR Watson   RA Jadeja  SL Malinga              0  ...   
179076    20     5  SN Thakur   RA Jadeja  SL Malinga              0  ...   
179077    20     6  SN Thakur   RA Jadeja  SL Malinga              0  ...   

        bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
0              0            0            0             0             0   
1              0            0            0             0             0   
2              0            0            0             0             4   
3              0            0            0             0             0   
4              0            0            0             0             0   
...          ...          ...          ...           ...           ...   
179073         0            0            0             0             1   
179074         0            0            0             0             2   
179075         0            0            0             0             1   
179076         0            0            0             0             2   
179077         0            0            0             0             0   

        extra_runs  total_runs  player_dismissed dismissal_kind    fielder  
0                0           0               NaN            NaN        NaN  
1                0           0               NaN            NaN        NaN  
2                0           4               NaN            NaN        NaN  
3                0           0               NaN            NaN        NaN  
4                2           2               NaN            NaN        NaN  
...            ...         ...               ...            ...        ...  
179073           0           1               NaN            NaN        NaN  
179074           0           2               NaN            NaN        NaN  
179075           0           1         SR Watson        run out  KH Pandya  
179076           0           2               NaN            NaN        NaN  
179077           0           0         SN Thakur            lbw        NaN  

[179078 rows x 21 columns]

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   season           756 non-null    int64 
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [8]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

In [9]:
df1.rename({"id":"match_id"},axis=1,inplace=True)

In [10]:
df=pd.merge(df2,df1,on='match_id')

In [11]:
df

match_id  inning         batting_team                 bowling_team  \
0              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
1              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
2              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
3              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
4              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
...          ...     ...                  ...                          ...   
179073     11415       2  Chennai Super Kings               Mumbai Indians   
179074     11415       2  Chennai Super Kings               Mumbai Indians   
179075     11415       2  Chennai Super Kings               Mumbai Indians   
179076     11415       2  Chennai Super Kings               Mumbai Indians   
179077     11415       2  Chennai Super Kings               Mumbai Indians   

        over  ball    batsman non_striker      bowler  is_super_over  ...  \
0          1     1  DA Warner    S Dhawan    TS Mills              0  ...   
1          1     2  DA Warner    S Dhawan    TS Mills              0  ...   
2          1     3  DA Warner    S Dhawan    TS Mills              0  ...   
3          1     4  DA Warner    S Dhawan    TS Mills              0  ...   
4          1     5  DA Warner    S Dhawan    TS Mills              0  ...   
...      ...   ...        ...         ...         ...            ...  ...   
179073    20     2  RA Jadeja   SR Watson  SL Malinga              0  ...   
179074    20     3  SR Watson   RA Jadeja  SL Malinga              0  ...   
179075    20     4  SR Watson   RA Jadeja  SL Malinga              0  ...   
179076    20     5  SN Thakur   RA Jadeja  SL Malinga              0  ...   
179077    20     6  SN Thakur   RA Jadeja  SL Malinga              0  ...   

        result  dl_applied               winner  win_by_runs  win_by_wickets  \
0       normal           0  Sunrisers Hyderabad           35               0   
1       normal           0  Sunrisers Hyderabad           35               0   
2       normal           0  Sunrisers Hyderabad           35               0   
3       normal           0  Sunrisers Hyderabad           35               0   
4       normal           0  Sunrisers Hyderabad           35               0   
...        ...         ...                  ...          ...             ...   
179073  normal           0       Mumbai Indians            1               0   
179074  normal           0       Mumbai Indians            1               0   
179075  normal           0       Mumbai Indians            1               0   
179076  normal           0       Mumbai Indians            1               0   
179077  normal           0       Mumbai Indians            1               0   

        player_of_match                                      venue  \
0          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
2          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
3          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
4          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
...                 ...                                        ...   
179073        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179074        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179075        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179076        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179077        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   

            umpire1    umpire2      umpire3  
0       AY Dandekar   NJ Llong          NaN  
1       AY Dandekar   NJ Llong          NaN  
2       AY Dandekar   NJ Llong          NaN  
3       AY Dandekar   NJ Llong          NaN  
4       AY Dandekar   NJ Llong          NaN  
...             ...        ...          ...  
179073  Nitin Menon  I

### The column which has the most missing values according to the data is:


In [12]:
(df.isna().sum())/len(df)*100

match_id             0.000000
inning               0.000000
batting_team         0.000000
bowling_team         0.000000
over                 0.000000
ball                 0.000000
batsman              0.000000
non_striker          0.000000
bowler               0.000000
is_super_over        0.000000
wide_runs            0.000000
bye_runs             0.000000
legbye_runs          0.000000
noball_runs          0.000000
penalty_runs         0.000000
batsman_runs         0.000000
extra_runs           0.000000
total_runs           0.000000
player_dismissed    95.066954
dismissal_kind      95.066954
fielder             96.399334
season               0.000000
city                 0.949307
date                 0.000000
team1                0.000000
team2                0.000000
toss_winner          0.000000
toss_decision        0.000000
result               0.000000
dl_applied           0.000000
winner               0.207731
win_by_runs          0.000000
win_by_wickets       0.000000
player_of_

**from the above the the fielder is the most missing**

In [13]:
df.drop(["player_dismissed","dismissal_kind","fielder","umpire3"],axis=1,inplace=True)

In [14]:
df["date"]=df["date"].astype("datetime64[ns, UTC]")

In [15]:
df["year"]=df["date"].dt.year

In [16]:
df["month"]=df["date"].dt.month

In [17]:
df["day"]=df["date"].dt.day

In [18]:
df1

match_id  season           city        date                        team1  \
0           1    2017      Hyderabad  2017-04-05          Sunrisers Hyderabad   
1           2    2017           Pune  2017-04-06               Mumbai Indians   
2           3    2017         Rajkot  2017-04-07                Gujarat Lions   
3           4    2017         Indore  2017-04-08       Rising Pune Supergiant   
4           5    2017      Bangalore  2017-04-08  Royal Challengers Bangalore   
..        ...     ...            ...         ...                          ...   
751     11347    2019         Mumbai    05/05/19        Kolkata Knight Riders   
752     11412    2019        Chennai    07/05/19          Chennai Super Kings   
753     11413    2019  Visakhapatnam    08/05/19          Sunrisers Hyderabad   
754     11414    2019  Visakhapatnam    10/05/19               Delhi Capitals   
755     11415    2019      Hyderabad    12/05/19               Mumbai Indians   

                           team2                  toss_winner toss_decision  \
0    Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1         Rising Pune Supergiant       Rising Pune Supergiant         field   
2          Kolkata Knight Riders        Kolkata Knight Riders         field   
3                Kings XI Punjab              Kings XI Punjab         field   
4               Delhi Daredevils  Royal Challengers Bangalore           bat   
..                           ...                          ...           ...   
751               Mumbai Indians               Mumbai Indians         field   
752               Mumbai Indians          Chennai Super Kings           bat   
753               Delhi Capitals               Delhi Capitals         field   
754          Chennai Super Kings          Chennai Super Kings         field   
755          Chennai Super Kings               Mumbai Indians           bat   

     result  dl_applied                       winner  win_by_runs  \
0    normal           0          Sunrisers Hyderabad           35   
1    normal           0       Rising Pune Supergiant            0   
2    normal           0        Kolkata Knight Riders            0   
3    normal           0              Kings XI Punjab            0   
4    normal           0  Royal Challengers Bangalore           15   
..      ...         ...                          ...          ...   
751  normal           0               Mumbai Indians            0   
752  normal           0               Mumbai Indians            0   
753  normal           0               Delhi Capitals            0   
754  normal           0          Chennai Super Kings            0   
755  normal           0               Mumbai Indians            1   

     win_by_wickets player_of_match  \
0                 0    Yuvraj Singh   
1                 7       SPD Smith   
2                10         CA Lynn   
3                 6      GJ Maxwell   
4                 0       KM Jadhav   
..              ...             ...   
751               9       HH Pandya   
752               6        AS Yadav   
753               2         RR Pant   
754               6    F du Plessis   
755               0       JJ Bumrah   

                                         venue         umpire1  \
0    Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1      Maharashtra Cricket Association Stadium  A Nand Kishore   
2       Saurashtra Cricket Association Stadium     Nitin Menon   
3                       Holkar Cricket Stadium    AK Chaudhary   
4                        M Chinnaswamy Stadium             NaN   
..                                         ...             ...   
751                           Wankhede Stadium   Nanda Kishore   
752                  M. A. Chidambaram Stadium     Nigel Llong   
753                           ACA-VDCA Stadium             NaN   
754                           ACA-VDCA Stadium   Sundaram Ravi   
755         Rajiv Gandhi Intl. Cricket Stadium     Nitin Me

In [44]:
df1["date"]=df1["date"].astype("datetime64[ns, UTC]")

In [45]:
df1["year"]=df1["date"].dt.year

In [46]:
df1["month"]=df1["date"].dt.month

In [47]:
df1["day"]=df1["date"].dt.day

In [48]:
df1

id  season           city                      date  \
0        1    2017      Hyderabad 2017-04-05 00:00:00+00:00   
1        2    2017           Pune 2017-04-06 00:00:00+00:00   
2        3    2017         Rajkot 2017-04-07 00:00:00+00:00   
3        4    2017         Indore 2017-04-08 00:00:00+00:00   
4        5    2017      Bangalore 2017-04-08 00:00:00+00:00   
..     ...     ...            ...                       ...   
751  11347    2019         Mumbai 2019-05-05 00:00:00+00:00   
752  11412    2019        Chennai 2019-07-05 00:00:00+00:00   
753  11413    2019  Visakhapatnam 2019-08-05 00:00:00+00:00   
754  11414    2019  Visakhapatnam 2019-10-05 00:00:00+00:00   
755  11415    2019      Hyderabad 2019-12-05 00:00:00+00:00   

                           team1                        team2  \
0            Sunrisers Hyderabad  Royal Challengers Bangalore   
1                 Mumbai Indians       Rising Pune Supergiant   
2                  Gujarat Lions        Kolkata Knight Riders   
3         Rising Pune Supergiant              Kings XI Punjab   
4    Royal Challengers Bangalore             Delhi Daredevils   
..                           ...                          ...   
751        Kolkata Knight Riders               Mumbai Indians   
752          Chennai Super Kings               Mumbai Indians   
753          Sunrisers Hyderabad               Delhi Capitals   
754               Delhi Capitals          Chennai Super Kings   
755               Mumbai Indians          Chennai Super Kings   

                     toss_winner toss_decision  result  dl_applied  ...  \
0    Royal Challengers Bangalore         field  normal           0  ...   
1         Rising Pune Supergiant         field  normal           0  ...   
2          Kolkata Knight Riders         field  normal           0  ...   
3                Kings XI Punjab         field  normal           0  ...   
4    Royal Challengers Bangalore           bat  normal           0  ...   
..                           ...           ...     ...         ...  ...   
751               Mumbai Indians         field  normal           0  ...   
752          Chennai Super Kings           bat  normal           0  ...   
753               Delhi Capitals         field  normal           0  ...   
754          Chennai Super Kings         field  normal           0  ...   
755               Mumbai Indians           bat  normal           0  ...   

    win_by_runs  win_by_wickets  player_of_match  \
0             1               0     Yuvraj Singh   
1             0               7        SPD Smith   
2             0              10          CA Lynn   
3             0               6       GJ Maxwell   
4             1               0        KM Jadhav   
..          ...             ...              ...   
751           0               9        HH Pandya   
752           0               6         AS Yadav   
753           0               2          RR Pant   
754           0               6     F du Plessis   
755           1               0        JJ Bumrah   

                                         venue         umpire1  \
0    Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1      Maharashtra Cricket Association Stadium  A Nand Kishore   
2       Saurashtra Cricket Association Stadium     Nitin Menon   
3                       Holkar Cricket Stadium    AK Chaudhary   
4                        M Chinnaswamy Stadium             NaN   
..                                         ...             ...   
751                           Wankhede Stadium   Nanda Kishore   
752                  M. A. Chidambaram Stadium     Nigel Llong   
753                           ACA-VDCA Stadium             NaN   
754                           ACA-VDCA Stadium   Sundaram Ravi   
755         Rajiv Gandhi Intl. Cricket Stadium     Nitin Menon   

            umpire2                  umpire3  year  month  day  
0          NJ Llong                      NaN  2017      4    5  
1            S Ravi  

### City which hosted most number of seasons according to the data is:


In [25]:
df1["city"].value_counts()

city
Mumbai            101
Kolkata            77
Delhi              74
Bangalore          66
Hyderabad          64
Chennai            57
Jaipur             47
Chandigarh         46
Pune               38
Durban             15
Bengaluru          14
Visakhapatnam      13
Centurion          12
Ahmedabad          12
Rajkot             10
Mohali             10
Indore              9
Dharamsala          9
Johannesburg        8
Cuttack             7
Ranchi              7
Port Elizabeth      7
Cape Town           7
Abu Dhabi           7
Sharjah             6
Raipur              6
Kochi               5
Kanpur              4
Nagpur              3
Kimberley           3
East London         3
Bloemfontein        2
Name: count, dtype: int64

**from the above mumbai is the most hosted city**

### Which year has the most number of matches played?

In [26]:
df1["year"].value_counts()

year
2013    76
2012    74
2011    73
2010    60
2014    60
2016    60
2018    60
2019    60
2017    59
2015    59
2008    58
2009    57
Name: count, dtype: int64

**in 2013 the most number of the matches are played**

In [27]:
cond=((df["winner"]=="Mumbai Indians")&(df["year"]==2011))

In [28]:
df[cond]

match_id  inning      batting_team           bowling_team  over  ball  \
56181       238       1  Delhi Daredevils         Mumbai Indians     1     1   
56182       238       1  Delhi Daredevils         Mumbai Indians     1     2   
56183       238       1  Delhi Daredevils         Mumbai Indians     1     3   
56184       238       1  Delhi Daredevils         Mumbai Indians     1     4   
56185       238       1  Delhi Daredevils         Mumbai Indians     1     5   
...         ...     ...               ...                    ...   ...   ...   
71968       305       2    Mumbai Indians  Kolkata Knight Riders    19     4   
71969       305       2    Mumbai Indians  Kolkata Knight Riders    19     5   
71970       305       2    Mumbai Indians  Kolkata Knight Riders    19     6   
71971       305       2    Mumbai Indians  Kolkata Knight Riders    20     1   
71972       305       2    Mumbai Indians  Kolkata Knight Riders    20     2   

               batsman      non_striker           bowler  is_super_over  ...  \
56181        DA Warner         V Sehwag  Harbhajan Singh              0  ...   
56182        DA Warner         V Sehwag  Harbhajan Singh              0  ...   
56183         V Sehwag        DA Warner  Harbhajan Singh              0  ...   
56184         V Sehwag        DA Warner  Harbhajan Singh              0  ...   
56185         V Sehwag        DA Warner  Harbhajan Singh              0  ...   
...                ...              ...              ...            ...  ...   
71968     JEC Franklin  Harbhajan Singh            B Lee              0  ...   
71969     JEC Franklin  Harbhajan Singh            B Lee              0  ...   
71970  Harbhajan Singh     JEC Franklin            B Lee              0  ...   
71971     JEC Franklin  Harbhajan Singh  Shakib Al Hasan              0  ...   
71972  Harbhajan Singh     JEC Franklin  Shakib Al Hasan              0  ...   

               winner  win_by_runs  win_by_wickets  player_of_match  \
56181  Mumbai Indians            0               8       SL Malinga   
56182  Mumbai Indians            0               8       SL Malinga   
56183  Mumbai Indians            0               8       SL Malinga   
56184  Mumbai Indians            0               8       SL Malinga   
56185  Mumbai Indians            0               8       SL Malinga   
...               ...          ...             ...              ...   
71968  Mumbai Indians            0               4         MM Patel   
71969  Mumbai Indians            0               4         MM Patel   
71970  Mumbai Indians            0               4         MM Patel   
71971  Mumbai Indians            0               4         MM Patel   
71972  Mumbai Indians            0               4         MM Patel   

                  venue    umpire1     umpire2  year  month day  
56181  Feroz Shah Kotla  AM Saheba   RB Tiffin  2011      4  10  
56182  Feroz Shah Kotla  AM Saheba   RB Tiffin  2011      4  10  
56183  Feroz Shah Kotla  AM Saheba   RB Tiffin  2011      4  10  
56184  Feroz Shah Kotla  AM Saheba   RB Tiffin  2011      4  10  
56185  Feroz Shah Kotla  AM Saheba   RB Tiffin  2011      4  10  
...                 ...        ...         ...   ...    ...  ..  
71968  Wankhede Stadium  Asad Rauf  SJA Taufel  2011      5  25  
71969  Wankhede Stadium  Asad Rauf  SJA Taufel  2011      5  25  
71970  Wankhede Stadium  Asad Rauf  SJA Taufel  2011      5  25  
71971  Wankhede Stadium  Asad Rauf  SJA Taufel  2011      5  25  
71972  Wankhede Stadium  Asad Rauf  SJA Taufel  2011      5  25  

[2416 rows x 37 columns]

In [30]:
df3=df[cond]

In [31]:
df3["winner"].unique()

array(['Mumbai Indians'], dtype=object)

In [32]:
len(df3["match_id"].unique())

10

**the most wins by the mumbai indian in 2011 are 10**

In [33]:
(df1["toss_decision"].value_counts())/len(df1)*100

toss_decision
field    61.243386
bat      38.756614
Name: count, dtype: float64

From the above 

the choose to field has taken 61.2

the choose to bat has taken 38.7566 

In [34]:
def abc(x):
    if x==0:
        return np.nan
    else:
        return 1

In [41]:
df1["win_by_runs"]=np.where((df1["win_by_runs"]==0),0,1)

In [42]:
df1["win_by_runs"].value_counts()

win_by_runs
0    419
1    337
Name: count, dtype: int64

In [50]:
df1["win_by_wickets"]=df1["win_by_wickets"].apply(abc)

In [51]:
df1[((df1["toss_decision"]=="field") & (df1["win_by_wickets"]==1))].groupby(by="year")["win_by_wickets"].sum()

year
2008    19.0
2009    13.0
2010    10.0
2011    27.0
2012    18.0
2013    14.0
2014    24.0
2015    13.0
2016    30.0
2017    28.0
2018    27.0
2019    30.0
Name: win_by_wickets, dtype: float64

### Most number of wickets taken by a bowler is


In [52]:
def dismissal_kind(x):
    if (x=="caught")|(x=="caught")|(x=="lbw")|(x=="caught and bowled")|(x=="stumped")|(x=="hit wicket"):
        return 1
    
    else:
        return 0

In [53]:
df2["Wickets"]=df2["dismissal_kind"].apply(dismissal_kind)

In [56]:
df3=df2.groupby(by="bowler")["Wickets"].sum().reset_index(name='wickets')



In [57]:
df3.sort_values(by='wickets')

bowler  wickets
164           K Goel        0
216        MB Parmar        0
220         MG Neser        0
78     CK Kapugedera        0
231    Mandeep Singh        0
..               ...      ...
264        PP Chawla      106
340       SL Malinga      107
126  Harbhajan Singh      121
93          DJ Bravo      127
7           A Mishra      131

[405 rows x 2 columns]

**from the above the A Mishra has taken the more no of wickets**

In [58]:
def batsman_runs(x):
    if (x==4)|(x==6):
        return 

In [59]:
df4=df2.groupby(by="batsman")["batsman_runs"].sum().reset_index(name="runs")

In [60]:
df4.sort_values(by="runs")

batsman  runs
55   Abdur Razzak     0
207       K Ahmed     0
373    RR Bhatkal     0
87        C Nanda     0
182       J Denly     0
..            ...   ...
392      S Dhawan  4632
112     DA Warner  4741
367     RG Sharma  4914
428      SK Raina  5415
486       V Kohli  5434

[516 rows x 2 columns]

### Bowlers with maximum number of extras:


In [61]:
df6=df2.groupby(by="bowler")["extra_runs"].sum().reset_index(name='extraruns')
df6

bowler  extraruns
0    A Ashish Reddy         14
1        A Chandila          3
2       A Choudhary          7
3       A Dananjaya          1
4        A Flintoff          1
..              ...        ...
400      YA Abdulla         17
401       YK Pathan         71
402       YS Chahal        116
403    Yuvraj Singh         27
404          Z Khan        169

[405 rows x 2 columns]

In [62]:
df6.sort_values(by="extraruns")

bowler  extraruns
357  Sachin Baby          0
77      CJ McKay          0
273      R Ninan          0
274      R Parag          0
70       C Munro          0
..           ...        ...
53       B Kumar        197
93      DJ Bravo        201
381     UT Yadav        219
251      P Kumar        236
340   SL Malinga        293

[405 rows x 2 columns]

**from the above the SL Malinga extra runs**

### Which venue has hosted most number of IPL matches?


In [63]:
df1["venue"].value_counts()

venue
Eden Gardens                                            77
M Chinnaswamy Stadium                                   73
Wankhede Stadium                                        73
Feroz Shah Kotla                                        67
Rajiv Gandhi International Stadium, Uppal               56
MA Chidambaram Stadium, Chepauk                         49
Sawai Mansingh Stadium                                  47
Punjab Cricket Association Stadium, Mohali              35
Maharashtra Cricket Association Stadium                 21
Subrata Roy Sahara Stadium                              17
Dr DY Patil Sports Academy                              17
Kingsmead                                               15
Punjab Cricket Association IS Bindra Stadium, Mohali    14
SuperSport Park                                         12
Sardar Patel Stadium, Motera                            12
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium     11
Brabourne Stadium                                 

### In 2017 when sunrisers hyderabad clashed against Royal Challengers Bangalore which team player won player of the match

In [64]:
cond=((df1["team1"]=="Sunrisers Hyderabad")&(df1["team2"]=="Royal Challengers Bangalore"))

In [65]:
df1[cond]

id  season       city                      date                team1  \
0        1    2017  Hyderabad 2017-04-05 00:00:00+00:00  Sunrisers Hyderabad   
431    432    2013  Bangalore 2013-04-09 00:00:00+00:00  Sunrisers Hyderabad   
480    481    2014  Bangalore 2014-05-04 00:00:00+00:00  Sunrisers Hyderabad   
567    568    2015  Hyderabad 2015-05-15 00:00:00+00:00  Sunrisers Hyderabad   
602    603    2016  Hyderabad 2016-04-30 00:00:00+00:00  Sunrisers Hyderabad   
635    636    2016  Bangalore 2016-05-29 00:00:00+00:00  Sunrisers Hyderabad   
674   7932    2018  Hyderabad 2018-07-05 00:00:00+00:00  Sunrisers Hyderabad   
706  11147    2019  Hyderabad 2019-03-31 00:00:00+00:00  Sunrisers Hyderabad   
749  11345    2019  Bengaluru 2019-04-05 00:00:00+00:00  Sunrisers Hyderabad   

                           team2                  toss_winner toss_decision  \
0    Royal Challengers Bangalore  Royal Challengers Bangalore         field   
431  Royal Challengers Bangalore          Sunrisers Hyderabad           bat   
480  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
567  Royal Challengers Bangalore          Sunrisers Hyderabad           bat   
602  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
635  Royal Challengers Bangalore          Sunrisers Hyderabad           bat   
674  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
706  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
749  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

     result  dl_applied  ... win_by_runs  win_by_wickets  player_of_match  \
0    normal           0  ...           1             NaN     Yuvraj Singh   
431  normal           0  ...           0             1.0          V Kohli   
480  normal           0  ...           0             1.0   AB de Villiers   
567  normal           1  ...           0             1.0          V Kohli   
602  normal           0  ...           1             NaN        DA Warner   
635  normal           0  ...           1             NaN      BCJ Cutting   
674  normal           0  ...           1             NaN    KS Williamson   
706  normal           0  ...           1             NaN       J Bairstow   
749  normal           0  ...           0             1.0        S Hetmyer   

                                         venue          umpire1  \
0    Rajiv Gandhi International Stadium, Uppal      AY Dandekar   
431                      M Chinnaswamy Stadium           S Ravi   
480                      M Chinnaswamy Stadium  HDPK Dharmasena   
567  Rajiv Gandhi International Stadium, Uppal     AK Chaudhary   
602  Rajiv Gandhi International Stadium, Uppal     AK Chaudhary   
635                      M Chinnaswamy Stadium  HDPK Dharmasena   
674  Rajiv Gandhi International Stadium, Uppal   Bruce Oxenford   
706         Rajiv Gandhi Intl. Cricket Stadium           S Ravi   
749                     M. Chinnaswamy Stadium      Nigel Llong   

                   umpire2        umpire3  year  month  day  
0                 NJ Llong            NaN  2017      4    5  
431             SJA Taufel            NaN  2013      4    9  
480            VA Kulkarni            NaN  2014      5    4  
567        HDPK Dharmasena            NaN  2015      5   15  
602        HDPK Dharmasena            NaN  2016      4   30  
635           BNJ Oxenford            NaN  2016      5   29  
674  Virender Kumar Sharma       O Nandan  2018      7    5  
706   KN Anantapadmanabhan  C Shamshuddin  2019      3   31  
749         Anil Chaudhary   Ulhas Gandhe  2019      4    5  

[9 rows x 21 columns]

from the above the player of the match is yuvaraj singh

### Across seasons who are the top three batsman's with most number of run out?

In [75]:
def dismissal_kind1(x):
    if x=="run out":
        return 1
    else:
        return 0

In [76]:
df2["runouts"]=df2["dismissal_kind"].apply(dismissal_kind1)

In [78]:
df3=df2.groupby(by="batsman")["runouts"].sum().reset_index(name='runouts')

In [79]:
df3

batsman  runouts
0    A Ashish Reddy        2
1        A Chandila        1
2          A Chopra        1
3       A Choudhary        1
4       A Dananjaya        0
..              ...      ...
511     YV Takawale        0
512   Yashpal Singh        0
513     Younis Khan        0
514    Yuvraj Singh        4
515          Z Khan        3

[516 rows x 2 columns]

In [80]:
df3.sort_values(by="runouts")

batsman  runouts
136      DS Lehmann        0
455  Sandeep Sharma        0
167   Harmeet Singh        0
168  Harpreet Singh        0
169      I Malhotra        0
..              ...      ...
486         V Kohli       15
147       G Gambhir       15
218      KD Karthik       16
367       RG Sharma       17
290        MS Dhoni       21

[516 rows x 2 columns]

### What are the total runs scored by V Kohli when the bowler was JJ Burmah?

In [81]:
cond1=((df2["batsman"]=="V Kohli")&(df2["bowler"]=='JJ Bumrah'))

In [82]:
df2[cond1].groupby(by="batsman")["batsman_runs"].sum()

batsman
V Kohli    112
Name: batsman_runs, dtype: int64

### Across all seasons which player was dismissed the maximum number of times via caught and bowled

In [83]:
def caughtbowled(x):
    if x=='caught and bowled':
        return 1
    else:
        return 0

In [84]:
df2["dismissal_kind"].unique()

array([nan, 'caught', 'bowled', 'run out', 'lbw', 'caught and bowled',
       'stumped', 'retired hurt', 'hit wicket', 'obstructing the field'],
      dtype=object)

In [95]:
df2["caught&bowled"]=df2["dismissal_kind"].apply(caughtbowled)

In [96]:
df3=df2.groupby(by="batsman")["caught&bowled"].sum().reset_index(name='caught&bowled')

In [98]:
df3.sort_values("caught&bowled")

batsman  caught&bowled
0    A Ashish Reddy              0
332  PD Collingwood              0
331     PC Valthaty              0
330        PA Reddy              0
328          P Shaw              0
..              ...            ...
53        AT Rayudu              5
42        AM Rahane              5
349        R Dravid              5
367       RG Sharma              6
428        SK Raina              7

[516 rows x 2 columns]

In [100]:
df2.groupby(by="batsman")["batsman_runs"].sum()

batsman
A Ashish Reddy     280
A Chandila           4
A Chopra            53
A Choudhary         25
A Dananjaya          4
                  ... 
YV Takawale        192
Yashpal Singh       47
Younis Khan          3
Yuvraj Singh      2765
Z Khan             117
Name: batsman_runs, Length: 516, dtype: int64